examples/contrib/detection/main.py

```diff
@@ -0,0 +1,329 @@
+
+    train_dataloader = idist.auto_dataloader(
+        train_dataset, batch_size=batch_size, shuffle=True, collate_fn=collate_fn, num_workers=4
+    )
+    val_dataloader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False, collate_fn=collate_fn, num_workers=4)
+    vis_dataloader = DataLoader(vis_dataset, batch_size=batch_size, shuffle=False, collate_fn=collate_fn, num_workers=4)
+
+    model = getattr(detection, model)(pretrained=True)
+    in_features = model.roi_heads.box_predictor.cls_score.in_features
+    model.roi_heads.box_predictor = FastRCNNPredictor(in_features, 21)
+    model = idist.auto_model(model)
+
+    scaler = GradScaler()
+    optimizer = SGD(lr=lr, params=model.parameters())
+    optimizer = idist.auto_optim(optimizer)
+    scheduler = OneCycleLR(optimizer, max_lr=lr, total_steps=len(train_dataloader) * epochs)
+
+    def update_model(engine, batch):
+        model.train()
+        images, targets = batch
+        images = list(image.to(device) for image in images)
+        targets = [{k: v.to(device) for k, v in t.items() if isinstance(v, torch.Tensor)} for t in targets]
+
+        with torch.autocast(device, enabled=True):
+            loss_dict = model(images, targets)
+            loss = sum(loss for loss in loss_dict.values())
+
+        optimizer.zero_grad()
+        scaler.scale(loss).backward()
+        scaler.step(optimizer)
+        scaler.update()
+
+        loss_items = {k: v.item() for k, v in loss_dict.items()}
+        loss_items["loss_average"] = loss.item() / 4
+
+        return loss_items
+
+    @torch.no_grad()
+    def inference(engine, batch):
+        model.eval()
+        images, targets = batch
+        images = list(image.to(device) for image in images)
+        outputs = model(images)
+        outputs = [{k: v.to("cpu") for k, v in t.items()} for t in outputs]
+        return {"y_pred": outputs, "y": targets, "x": [i.cpu() for i in images]}
+
+    trainer = Engine(update_model)
+    evaluator = Engine(inference)
+    visualizer = Engine(inference)
+
+    aim_logger = AimLogger(
+        repo=os.path.join(log_dir, "aim"),
+        experiment=experiment_name,
+    )
+
+    CocoMetric(convert_to_coco_api(val_dataset)).attach(evaluator, "mAP")
+
+    @trainer.on(Events.EPOCH_COMPLETED)
+    @one_rank_only()
+    def log_validation_results(engine):
+        evaluator.run(val_dataloader)
+        visualizer.run(vis_dataloader)
+
+    @trainer.on(Events.ITERATION_COMPLETED)
+    def step_scheduler(engine):
+        scheduler.step()
+        aim_logger.log_metrics({"lr": scheduler.get_last_lr()[0]}, step=engine.state.iteration)
+
+    @visualizer.on(Events.EPOCH_STARTED)
+    def reset_vis_images(engine):
+        engine.state.model_outputs = []
+
+    @visualizer.on(Events.ITERATION_COMPLETED)
+    def add_vis_images(engine):
+        engine.state.model_outputs.append(engine.state.output)
+
+    @visualizer.on(Events.ITERATION_COMPLETED)
+    def submit_vis_images(engine):
```

Comment:
Does this ok in distributed ? I would have added a `@one_rank_only()` decorator.

Reply:
I havn't tested it on multi gpu core yet, there is a `@one_rank_only()` protection on the `trainer`'s `epoch_end` so the visualizer engine will not be called on other ranks.